fix(ivy): warn instead of throwing for unknown elements

crisbeto opened the pull request:
Follow-up from [this discussion](https://github.com/angular/angular/pull/33419#discussion_r339296216). In Ivy we don't use the schema to validate tag names, but instead we use feature detection to figure out whether an element is supported. While this should generally be more accurate, it'll also end up throwing for some more innocent cases. E.g. now Ivy throws an error for `main` elements in IE which is accurate since IE doesn't support the element, but is annoying since there is no functionality attached.

These changes switch to logging a warning instead, similarly to what we're doing for unknown properties.

kara commented on packages/core/test/acceptance/ng_module_spec.ts:
```diff
@@ -166,36 +166,69 @@ describe('NgModule', () => {
-           fixture.detectChanges();
-         }).toThrowError(/'custom' is not a known element/);
-       });
+    onlyInIvy('unknown element check logs a warning rather an throwing')
```
```suggestion
    onlyInIvy('unknown element check logs a warning rather than throwing')
```

kara commented on packages/core/test/acceptance/ng_module_spec.ts:
```diff
@@ -166,36 +166,69 @@ describe('NgModule', () => {
+              expect(spy.calls.mostRecent().args[0]).toMatch(/'custom-el' is not a known element/);
+            });
+
+    modifiedInIvy('unknown element error thrown instead warning')
```
Same below

```suggestion
    modifiedInIvy('unknown element error thrown instead of warning')
```

kara commented on packages/core/test/acceptance/ng_module_spec.ts:
```diff
@@ -166,36 +166,69 @@ describe('NgModule', () => {
     onlyInIvy('test relies on Ivy-specific AOT format')
-        .it('should not throw unknown element error for AOT-compiled components', () => {
+        .it('should not log unknown element warning for AOT-compiled components', () => {
+          const spy = spyOn(console, 'warn');
```
I think we lost the `modifiedInIvy` version of this test (unless I'm missing it)? Let's add one

kara commented on packages/core/test/acceptance/ng_module_spec.ts:
```diff
@@ -253,88 +286,173 @@ describe('NgModule', () => {
+        });
+
+    onlyInIvy('unknown element check logs a warning rather than throwing')
+        .it('should not throw unknown element error with CUSTOM_ELEMENTS_SCHEMA', () => {
```
```suggestion
        .it('should not warn about unknown elements with CUSTOM_ELEMENTS_SCHEMA', () => {
```

kara commented on packages/core/test/acceptance/ng_module_spec.ts:
```diff
@@ -253,88 +286,173 @@ describe('NgModule', () => {
-      class MyComp {
-      }
+    onlyInIvy('unknown element check logs a warning rather than throwing')
+        .it('should not throw unknown element error with NO_ERRORS_SCHEMA', () => {
```
```suggestion
        .it('should not warn about unknown elements with NO_ERRORS_SCHEMA', () => {
```

kara commented on packages/core/test/acceptance/ng_module_spec.ts:
```diff
@@ -253,88 +286,173 @@ describe('NgModule', () => {
-      class CustomEl {
-      }
+    onlyInIvy('unknown element check logs a warning rather than throwing')
+        .it('should not throw unknown element error if element matches a directive', () => {
```
```suggestion
        .it('should not warn about unknown elements if element matches a directive', () => {
```

kara commented on packages/core/test/acceptance/ng_module_spec.ts:
```diff
@@ -253,88 +286,173 @@ describe('NgModule', () => {
+        });
+
+    onlyInIvy('unknown element check logs a warning rather than throwing')
+        .it('should not throw for HTML elements inside an SVG foreignObject', () => {
```
```suggestion
        .it('should not warn for HTML elements inside an SVG foreignObject', () => {
```

kara reviewed (changes requested):
Mostly looks good, just need to update the Ivy-only test descriptions

crisbeto committed:
fix(ivy): warn instead of throwing for unknown elements

Follow-up from [this discussion](https://github.com/angular/angular/pull/33419#discussion_r339296216). In Ivy we don't use the schema to validate tag names, but instead we use feature detection to figure out whether an element is supported. While this should generally be more accurate, it'll also end up throwing for some more innocent cases. E.g. now Ivy throws an error for `main` elements in IE which is accurate since IE doesn't support the element, but is annoying since there is no functionality attached.

These changes switch to logging a warning instead, similarly to what we're doing for unknown properties.

crisbeto commented:
The feedback has been addressed.

AndrewKushnir commented:
[Presubmit](http://test/OCL:288523430:BASE:288523461:1578421553445:c090ba54)

kara reviewed (approved):
LGTM

kara commented:
Caretaker note: This PR requires a patch [here](https://cs.corp.google.com/piper///depot/google3/javascript/angular2/testing/catalyst/integration_test.ts?l=144&ws=tap-prod-presubmit/58690893&snapshot=2) because the project is explicitly testing that the framework throws an error.

AndrewKushnir commented:
Thanks @kara. I've created a [patch](http://cl/288543848) that disables that test temporary. The patch can be submitted prior to landing this PR.
